File: network/netdiag_v1/internal/schema/schema.go
// Package schema defines the versioned snapshot envelope shared by every
// collector and, later, by the machine diagnostician (parent spec v5 §4.3).
package schema

import "time"

// 0.2.0: v1 adds fact keys (new collectors) — envelope unchanged, minor bump.
const SchemaVersion = "0.2.0"

// Status values for the per-collector envelope. Absence is never health:
// a collector that did not run reports skipped/timeout/error with a reason.
const (
	StatusOK      = "ok"
	StatusSkipped = "skipped"
	StatusTimeout = "timeout"
	StatusError   = "error"
)

// Privilege levels (spec §3.1).
const (
	PrivUnprivileged = "unprivileged"
	PrivElevated     = "elevated"
)

// CollectorResult is the per-collector envelope (spec §4).
type CollectorResult struct {
	Status         string         `json:"status"`
	Reason         string         `json:"reason,omitempty"`
	DurationMS     int64          `json:"duration_ms"`
	PrivilegeLevel string         `json:"privilege_level"`
	Data           map[string]any `json:"data,omitempty"`
}

// Snapshot is the single structured blob (spec §4.3) — same outer shape
// regardless of OS, versioned from day one.
type Snapshot struct {
	SchemaVersion string                     `json:"schema_version"`
	Tool          string                     `json:"tool"`
	ToolVersion   string                     `json:"tool_version"`
	CollectedAt   time.Time                  `json:"collected_at"`
	Hostname      string                     `json:"hostname"`
	OS            string                     `json:"os"`
	Collectors    map[string]CollectorResult `json:"collectors"`
}

// Facts flattens every ok collector's Data into one namespace for the
// rules engine. Collector authors keep keys globally unique on purpose.
func (s *Snapshot) Facts() map[string]any {
	facts := map[string]any{}
	for _, c := range s.Collectors {
		if c.Status != StatusOK {
			continue
		}
		for k, v := range c.Data {
			facts[k] = v
		}
	}
	return facts
}

File: diagnostics/diagnostic-companion-v1.3/diagnostic-companion/schema.py
"""
Shared snapshot schema (spec §4.3).

A snapshot is one JSON-serialisable dict:

{
  "schema_version": "0.1.0",
  "collected_at": "2026-07-20T10:00:00Z",
  "hostname": "...",
  "os": "linux" | "windows",
  "sections": {
    "<collector_id>": {
      "status": "ok" | "skipped" | "timeout" | "error",
      "reason": str | None,
      "duration_ms": int,
      "privilege_level": "unprivileged" | "elevated",
      "data": {...}
    },
    ...
  }
}

Every collector's output MUST go through build_envelope() so that
"absence is never health" (spec §3.4) is mechanically enforced rather
than left to each collector to remember.
"""

SCHEMA_VERSION = "0.1.0"

VALID_STATUSES = {"ok", "skipped", "timeout", "error"}


def build_envelope(status, reason, duration_ms, privilege_level, data):
    if status not in VALID_STATUSES:
        raise ValueError(f"invalid status: {status!r}")
    return {
        "status": status,
        "reason": reason,
        "duration_ms": duration_ms,
        "privilege_level": privilege_level,
        "data": data or {},
    }
